core/tests/Test.Microsoft.Identity.Core.Unit/Telemetry/TelemetryHelperTests.cs

```diff
@@ -0,0 +1,173 @@
+﻿//----------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using Microsoft.Identity.Core;
+using Microsoft.Identity.Core.Telemetry;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Test.Microsoft.Identity.Core.Unit.Telemetry
+{
+    [TestClass]
+    public class TelemetryHelperTests
+    {
+        private const string requestId = "therequestid";
+        private _TestTelem _telem;
+        private _TestEvent _startEvent;
+        private _TestEvent _stopEvent;
+
+        [TestInitialize]
+        public void Setup()
+        {
+            _telem = new _TestTelem();
+            CoreTelemetryService.InitializeCoreTelemetryService(_telem);
+
+            _startEvent = new _TestEvent("start event");
+            _stopEvent = new _TestEvent("stop event");
+        }
+
+        [TestCleanup]
+        public void Cleanup()
+        {
+            _telem = null;
+            CoreTelemetryService.InitializeCoreTelemetryService(null);
+        }
+
+        private class _TestEvent : EventBase
+        {
+            public _TestEvent(string eventName) : base(eventName)
+            {
+            }
+        }
+
+        private class _TestTelem : ITelemetry

```

Comment:
We have a dependency on NSubstitute that we use to to mock objects, count calls etc. It makes the code a bit more succinct, although creating the mock manually does make debugging tests a lot easier. Just a thought / suggestion, this is perfectly fine and easy to read as is.

Reply:
Yeah, I saw that dependency and its usage with the Logger infra.  I've used Moq in the past which is super helpful but we didn't have a dependency on it yet.  Is NSubstitute a full mocking framework like Moq?  If so I'd rather use something like that.  I've just only used Moq previously...

Reply:
Yeah, afaik NSubstitute does everything that Moq does but plays a bit better with C#. The important thing is that they both target netstandard1.3, which means we can use these libraries for all target framework tests (except ADAL netstandard1.1 which we should probably deprecate sooner or later). Currently we have only net45 tests, but I have a work item that enables other tests as well.